[tests] JavaScript: extract searchindex.js-format test fixtures.

jayaddison opened the pull request:
### Feature or Bugfix
- Refactoring

### Purpose
- Reformats the test `index` data in the JavaScript HTML search tests to match the format emitted by the Sphinx (Python) HTML builder.

### Detail
- The content of `searchindex.js` is a JavaScript code file that runs `Search.setIndex(...)` with the contents of the index as an argument.
- Instead of loading the index per-test from a variable, retrieve each fixture from the built-in local `karma` HTTP server and `eval` it.
- Includes fixtures generated from genuine, buildable test Sphinx project sources included within this repository.
- Fails the Python test suite when regenerated `searchindex.js` fixture content does not match the existing fixture files.

### Relates
- Resolves #12099.

jayaddison committed:
[tests] JavaScript: extract searchindex.js-format test fixtures

jayaddison committed:
[tests] JavaScript: write Sphinx sources that produce acceptable searchindex.js test fixtures

jayaddison committed:
[tests] linting: add a ruff exclude filter for the JS fixture source roots

jayaddison committed:
[tests] nitpick: reduce the changeset diff by retaining an index variable in test cases

jayaddison committed:
Add CHANGES.rst entry

jayaddison commented on tests/js/searchtools.js:
```diff
@@ -1,16 +1,19 @@
-        titleterms:{}
-      }
-      Search.setIndex(index);
+      searchindex = loadFixture("index.cpp.js");
```
Ah, hm.  I have a suggestion for myself here: load these using the filename `searchindex.js`, but from a directory that has the same name as the source directory under `js/roots`.

That'll be clearer and easier when mapping the source artifacts to their destinations.

jayaddison committed:
[tests] fixup: regenerate multiterm test search index

jayaddison committed:
[tests] refactor: use a separate index fixture directory per root

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
[search] Refactor search code to improve testability

jayaddison committed:
[tests] expect duplicate result due to #11961

picnixz commented on tests/js/fixtures/multiterm/searchindex.js:
```diff
@@ -0,0 +1 @@
+Search.setIndex({"alltitles": {"Main Page": [[0, "main-page"]]}, "docnames": ["index"], "envversion": {"sphinx": 61, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2}, "filename ...
```
Since it's a fixture, could you make it larger? like more readable? (prettify it). It doesn't hurt for debugging / extending the fixture.

jayaddison replied:
Yep, I was thinking about doing this too.  Especially if/when the format of the index changes, that could be very useful.

To do that, we'd want to edit the Python [`_JavaScriptIndex.dumps` code](https://github.com/sphinx-doc/sphinx/blob/6aaeee206bda210f1b2cd845a924f86b18c0af8e/sphinx/search/__init__.py#L161-L165) to set `indent=4` or something similar.  But we wouldn't want that by default for most projects, because it'll increase the file size of the index.

jayaddison replied:
I think I'd file a feature request for this: a config setting `html_search_minify_index` (or similar), default `True`.  We would set it to `False` in the configuration files in the JS test roots.

picnixz replied:
Wait, the fixtures are always auto-generated? if so, you can just run a post-processing step where you prettify that specific fixture, right?

jayaddison replied:
That could be an easier option, yes!

jayaddison replied:
The generation occurs within a Python context (running the Sphinx build) so I am considering https://github.com/beautifier/js-beautify which can be installed using `pip`.

jayaddison replied:
(it would be a post-processing step, but even so, we'll want something that can easily run on any developer/CI machine, preferably without installing system-level packages)

picnixz replied:
Ok, for now let's leave it like this

jayaddison replied:
Ah, I missed that comment before developing and pushing the change.

jayaddison replied:
> Wait, the fixtures are always auto-generated? if so, you can just run a post-processing step where you prettify that specific fixture, right?

I'm going to remove the post-processing again I think, and there's a specific reason I have in mind: by doing that, I can add a Python test that builds each of these roots and confirms that the `searchindex.js` in the JS test directory is correct.

They should change extremely rarely, and if people want to view prettified diffs of those, they can do relatively easily locally.

picnixz commented on tests/js/fixtures/cpp/searchindex.js:
```diff
@@ -0,0 +1 @@
+Search.setIndex({"alltitles": {}, "docnames": ["index"], "envversion": {"sphinx": 61, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2}, "filenames": ["index.rst"], "indexentrie ...
```
Same comment as above.

picnixz commented on tests/js/searchtools.js:
```diff
@@ -1,16 +1,19 @@
 describe('Basic html theme search', function() {
 
+  function loadFixture(name) {
```
I never used karma but should I assume that this is the way to do it? namely doing an XMLHttpRequest? isn't it possible just to like.. load the file normally?

jayaddison replied:
Good question.  So, two considerations: `karma` imports all configured JavaScript files when the tests run, and the `searchindex.js` files affect the state of the application (because it calls `setIndex`).

We want each test case to call `setIndex` individually -- setting state only for that test.

Doing that requires opening and `eval(...)`'ing the fixture.  The tests run in a browser, so they can't open local files -- and so we use the `karma` HTTP server to serve it up for access instead.

picnixz replied:
Oh I see. Good then

jayaddison replied:
Thank you.

picnixz commented on tests/js/searchtools.js:
```diff
@@ -1,16 +1,19 @@
-      }
-      Search.setIndex(index);
+      searchindex = loadFixture("cpp/searchindex.js");
+      eval(searchindex);
```
You could perhaps create another function called `evalFixture` for that? or is there a more canonical way to do it with karma?

jayaddison replied:
I'll double-check the documentation; perhaps there is a better way (it would seem like importing a script file per-test could be a common use case).

jayaddison replied:
I didn't find anything suitable in `karma` itself.  I did find some plugins.. there's [a reasonably popular one](https://github.com/billtrik/karma-fixture) that loads HTML and JSON, but I don't think it is keen on [evaluating JavaScript](https://github.com/billtrik/karma-fixture/issues/19).

I had a look at `jasmine`, too.  We could enable a plugin to provide [mocking of AJAX requests](https://jasmine.github.io/tutorials/mocking_ajax) -- but again given that the code is running in a browser context, then we need to read the fixture data from somewhere.  I'll think about this some more.

jayaddison committed:
[tests] run 'js-beautify -r ...' on each searchindex.js file

jayaddison committed:
[tests] Fixup: regenerate cpp searchindex.js

jayaddison committed:
Merge branch 'issue-12099-prep/extract-js-search-index-fixtures' into refactor/extract-search-result-gathering

jayaddison committed:
[utils] Add script to regenerate JavaScript test fixtures

jayaddison committed:
[utils] Add script to regenerate JavaScript test fixtures

jayaddison committed:
[tests] nitpick: relocate project-related JavaScript files away from directory containing test case files

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'issue-12099-utils/fixture-regeneration' into issue-12099-prep/extract-js-search-index-fixtures

Conflicts:
	sphinx/themes/basic/static/searchtools.js

jayaddison committed:
[tests] Regenerate test fixtures by running 'python utils/generate_js_fixtures.py'

jayaddison committed:
Merge branch 'issue-12099-prep/extract-js-search-index-fixtures' into refactor/extract-search-result-gathering

jayaddison committed:
[utils] Remove beautification step for generated searchindex.js files.

Ref: https://github.com/sphinx-doc/sphinx/pull/12102#discussion_r1535606662

jayaddison committed:
Merge branch 'issue-12099-utils/fixture-regeneration' into refactor/extract-search-result-gathering

Conflicts:
	utils/generate_js_fixtures.py

jayaddison committed:
[tests] Regenerate searchindex.js files.

jayaddison committed:
Merge branch 'issue-12099-utils/fixture-regeneration' into issue-12099-prep/extract-js-search-index-fixtures

Conflicts:
	utils/generate_js_fixtures.py

jayaddison committed:
[linting] Reformat using 'ruff format'.

(cherry picked from commit 9bed49541ec660420273b2e02724b15d09c9fd07)

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
[tests] Regenerate searchindex.js files.

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'refactor/extract-search-result-gathering' into issue-12099-prep/extract-js-search-index-fixtures

Conflicts:
	utils/generate_js_fixtures.py

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
[utils] Cleanup: remove unused function.

Relates-to commit a74bc156d5f10ce3e5f87cfc56f51d7100976e9b.

jayaddison committed:
[tests] Add test case to confirm that JavaScript search fixtures remain the same when regenerated.

jayaddison committed:
[docs] Add debugging tip for JavaScript searchindex.js fixtures.

jayaddison committed:
[tests] Refactor: rewrite searchindex test using pytest parametrize.

jayaddison committed:
[tests] Refactor: use temporary directory when regenerating search indexes.

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Revert "[tests] nitpick: relocate project-related JavaScript files away from directory containing test case files"

This reverts commit d619a5211a5b863896996e71ee92ffb7f60866d5.

jayaddison committed:
[tests] karma config: explicitly add language_data.js alongside documentation_options.js entry.

jayaddison committed:
[tests] nitpick: relocate file-read operations to same-code-line as hashing operation.

jayaddison committed:
[tests] nitpick / consistency: use current-file as base for both project and fixture path lookups.

jayaddison committed:
[tests] refactor / cleanup: re-use existing TESTS_ROOT testutil constant.

jayaddison committed:
[tests] search: indicate full path of search index file when it fails a regeneration check.

jayaddison committed:
[docs] fixup for JS search-index fixture paths.

jayaddison committed:
[docs] contributing: rewrite paragraph about JavaScript test fixtures.

jayaddison committed:
[docs] contributing: rewrite (again) paragraph about JavaScript test fixtures.

jayaddison committed:
[tests] karma config: nitpick: move served files to top of files option.

jayaddison committed:
[search] nitpick: remove empty line from query function.

picnixz commented on tests/js/language_data.js:
```diff
@@ -0,0 +1,26 @@
+var stopwords = [];
+
+
+/* Non-minified version is copied as a separate JS file, is available */
```
```suggestion
/* Non-minified version is copied as a separate JS file, if available */
```

Is it a typo?

jayaddison replied:
Does seem to be, yep - thanks!

picnixz commented on tests/js/searchtools.js:
```diff
@@ -24,34 +22,34 @@ describe('Basic html theme search', function() {
-        15,
-        'index.rst']];
-      expect(Search.performTermsSearch(searchterms, excluded, terms, titleterms)).toEqual(hits);
+      eval(loadFixture("multiterm/searchindex.js"));
```
I don't understand why the two tests are now different.

jayaddison replied:
The closest analogy for this that I can think of is the `srcdir=...` setting for the Python tests.  Each of the JavaScript test cases can (and in the case of the search tests, does) now read from test fixture(s) relevant to that individual test case.

It might be possible to refactor this so that the `loadFixture` happens implicitly / during some kind of `setup` phase based on the test name.

picnixz replied:
Actually, I don't understand why the duplicated are found with the fixture but not with the previous test (was there anything that changed? I thought that you just moved the hard-coded fixtures in a separate fixture file).

jayaddison replied:
> It might be possible to refactor this so that the `loadFixture` happens implicitly / during some kind of `setup` phase based on the test name.

The relevant `jasmine` guidance to follow if/when doing this appears to be: https://jasmine.github.io/tutorials/sharing_behaviors#sharing-behaviors

jayaddison replied:
> Actually, I don't understand why the duplicated are found with the fixture but not with the previous test (was there anything that changed? I thought that you just moved the hard-coded fixtures in a separate fixture file).

Good question.  Checking why that happened..

jayaddison replied:
Ah yes, I forgot about this.  The old test code was only testing a portion of the search query handling - the `performTermsSearch` function.

The duplicate is found during another part of the search code, namely here: https://github.com/sphinx-doc/sphinx/blob/600d3d3adaee530363c88ab1a000ef83e5db428e/sphinx/themes/basic/static/searchtools.js#L323-L338

(note when navigating further down that code file that [`performTermsSearch` is called later](https://github.com/sphinx-doc/sphinx/blob/600d3d3adaee530363c88ab1a000ef83e5db428e/sphinx/themes/basic/static/searchtools.js#L367-L368))

When we update the code here to use the top-level `_parseQuery` function -- that encompasses all of the query-handling code -- the duplicate appears in the results because we check both `term` and `title` matches.

picnixz replied:
Could you make the old test compatible with your fixtures as well? it'd be good to test specific components.

jayaddison replied:
To check I understand you: partially revert the changes to the `should be able to search for multiple terms`, so that it continues to load a search index from fixtures, and only tests the `performTermsSearch` function, as-before.

picnixz replied:
With the testroot in Python you have E2E but sometimes you only want to check a single component's logic so I would like to keep single component logic testing separate. In other words:

- keep the old test (manually creating the fixture is fine); it's just a logic unit being tested.
- you create a new test for testing what you just added (performing the full E2E, with the duplicates).

jayaddison replied:
Makes sense - I'm thinking about it; doing that would mean that the two tests duplicate a lot of the coverage and behaviour that they're testing.

I'm also wondering whether to extract the query refactoring into a separate smaller pull request, without the test fixture changes.

jayaddison replied:
> I'm also wondering whether to extract the query refactoring into a separate smaller pull request, without the test fixture changes.

Yep, let's do that first, if that's OK?  I've opened #12229 for it.

jayaddison replied:
> > With the testroot in Python you have E2E but sometimes you only want to check a single component's logic so I would like to keep single component logic testing separate. In other words:
> >
> >  * keep the old test (manually creating the fixture is fine); it's just a logic unit being tested.
> >
> >  * you create a new test for testing what you just added (performing the full E2E, with the duplicates).
> 
> Makes sense - I'm thinking about it; doing that would mean that the two tests duplicate a lot of the coverage and behaviour that they're testing.

Ok, agreed - I think it is OK to have both types of test in place, and I've pushed corresponding changes.

picnixz commented on utils/generate_js_fixtures.py:
```diff
@@ -0,0 +1,26 @@
+
+
+def build(srcdir: Path) -> None:
+    cmd = ('sphinx-build', '-E', '-q', '-b', 'html', f'{srcdir}', f'{srcdir}/build')
```
Should we really make it in build or in _build instead?

jayaddison replied:
Probably better to use `_build` indeed.  Updated (including the `.gitignore` file.  there may be some unrelated stale entries in there to clean up separately).

picnixz commented on utils/generate_js_fixtures.py:
```diff
@@ -0,0 +1,26 @@
+import subprocess
+from pathlib import Path
+
+SPHINX_ROOT = Path(__file__).resolve().parents[1]
```
```suggestion
SPHINX_ROOT = Path(__file__).resolve().parent.parent
```

(I think it's the correct one, but check it with a print).

jayaddison replied:
That does work, yep - it's less consistent with other scripts in the same directory, so I'm unsure whether to change it.

picnixz replied:
I updated my suggestion since it's better to first resolve and do the parent.parent (in case you have symlinks). It's maybe less consistent but it's clearer. I mean, you need to remember that your parents are in the reverse order whereas with .parent.parent it's clearer to me.

jayaddison committed:
[search] typo fixup: 'if' instead of 'is'.

jayaddison committed:
[utils] js fixtures: use '_build' as output dir instead of 'build'.

jayaddison committed:
[utils] update path navigation steps during JavaScript fixture generation code.

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

jayaddison committed:
[tests] search: re-constrain test coverage for multiterm test case to exclude titles.

jayaddison committed:
[search] Rectification: rename misleadingly-named 'searchTerms' variable to 'searchParameters'.

(cherry picked from commit 7c151e4cf9f0bca8a08fe459a67f547b6ae9f878)

jayaddison committed:
[search] Rectification: test cases: rename misleadingly-named 'searchTerms' variable to 'searchParameters'.

Relates-to commit 2ef85babf3e96b8266a955cdd52d2416651013a9.

jayaddison committed:
[tests] search: add end-to-end / search-result-aggregation test.

jayaddison committed:
[tests] search: fixup: initialize index-term variables from Search._index

The Search._index property is assigned during the setIndex(...) function call, as performed by our evaluation of searchindex.js content.

Fixup-for / relates-to commit 7370d2056d61225a47d50ac279bbb1a12caa6227.

jayaddison committed:
[tests] search: migrate partial-match search to use fixture data.

Post-merge / relates-to commit 94b9fc5698d5f7216044783d14113c35efcbfeec.

jayaddison committed:
[tests] search: re-constrain cpp and partial test cases to term-search-only.

Follows-on-from commit 7370d2056d61225a47d50ac279bbb1a12caa6227.

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison commented on tests/js/searchtools.js:
```diff
@@ -1,20 +1,21 @@
-        titleterms:{}
-      }
-      Search.setIndex(index);
+      eval(loadFixture("cpp/searchindex.js"));
```
This test case has begun failing, and the reason is that the token `C++` (as found in a source document) is typically reduced down to the single-character term `c` in the search index.

I think it's best to resolve this by using `Search.parseQuery` to form a query token that will match `c`, instead of trying to get a literal term `c++` into the index (which is possible, but perhaps not best practice).

jayaddison committed:
[tests] search: use Search.parseQuery function to match query terms to indexed terms.

jayaddison commented on tests/test_search.py:
```diff
@@ -341,3 +344,20 @@ def assert_is_sorted(item, path):
+    fresh_hash = sha256(fresh_searchindex.read_bytes()).digest()
+
+    existing_searchindex = (TESTS_ROOT / 'js' / 'fixtures' / directory.name / 'searchindex.js')
+    existing_hash = sha256(existing_searchindex.read_bytes()).digest()
```
Some caution about these lines: I _thought_ that using `pathlib.Path` files' `read_bytes()` and hashing the results of that should produce identical results on different platforms.

A [test failure](https://github.com/sphinx-doc/sphinx/actions/runs/8563318800/job/23468161852#step:6:2342) on Windows where differences appear with very similar code (from an otherwise-unrelated PR) is making me want to double-check that.

jayaddison replied:
Ok, in the case of `searchindex.js` this appears not to be a problem - I believe due to the absence of newlines in the file content (meaning that the byte-for-byte content of the file is unchanged when written on different operating system platforms).

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
[utils] js fixtures: create destination parent directory if it does not already exist.

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

Conflicts:
	tests/test_search.py

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

wlach commented on CHANGES.rst:
```diff
@@ -313,6 +313,7 @@ Testing
   As a side-effect, this removes the need for test server lockfiles,
   meaning that any remaining ``tests/test-server.lock`` files can safely be
   deleted.
+* karma: refactor HTML search tests to use fixtures generated by Sphinx.
```
This needs to be moved up to the current release. Don't forget to credit yourself!

wlach commented on tests/js/language_data.js:
```diff
@@ -0,0 +1,26 @@
+ * ~~~~~~~~~~~~~~~~
+ *
+ * This script contains the language-specific data used by searchtools.js,
+ * namely the list of stopwords, stemmer, scorer and splitter.
```
The comments in this file seem a little misleading? This is really just test scaffolding AFAICT

jayaddison replied:
Mm, yep - this is an interesting case - it's an evaluated copy of a [template file](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/themes/basic/static/language_data.js_t) from the basic theme.  That's me attempting to make the JavaScript test environment realistic, but it is duplicative (and a bit confusing).

wlach commented on tests/js/roots/multiterm/index.rst:
```diff
@@ -0,0 +1,4 @@
+Main Page
+=========
+
+Welcome to the... main page!
```
I would personally use this opportunity to add a bit of context about what this file tests.

wlach commented on tests/js/roots/partial/index.rst:
```diff
@@ -0,0 +1,4 @@
+sphinx_utils module
+===================
+
+Useful utilities.
```
Same, I would personally use this opportunity to add a bit of context about what this file tests.

wlach commented on tests/js/searchtools.js:
```diff
@@ -1,20 +1,20 @@
+  function loadFixture(name) {
+      req = new XMLHttpRequest();
+      req.open("GET", `base/tests/js/fixtures/${name}`, false);
+      req.send(null);
```
I'm guessing you used `XMLHttpRequest()` because you want to make a synchronous? Should probably add a comment about that.

Apparently using [XMLHttpRequest this way is discouraged](https://stackoverflow.com/a/72561702/295132) and might be deprecated at some point, but they've been saying that since 2014. 😉 It's probably fine.

jayaddison replied:
Ok, that's good to know :)  Yep; we don't want the test to proceed until the fixture has definitely loaded -- I'm inclined to use a straightforward (even if seemingly dated) approach here.

wlach commented on tests/js/searchtools.js:
```diff
@@ -55,18 +44,11 @@ describe('Basic html theme search', function() {
-      titleterms = index.titleterms;
+      eval(loadFixture("partial/searchindex.js"));
+
+      [/* first-ignored */, searchterms, excluded, /*rest-ignored*/] = Search._parseQuery('sphinx');
```
I would personally use underscores in these cases but it's not that big a deal.

```suggestion
      [_, searchterms, excluded, _] = Search._parseQuery('sphinx');
```

jayaddison replied:
JavaScript's unpacking behaviour here seems somewhat inconsistent to me; the value on the left is a single item from the value on the right-hand-side, and yet the last value (the second `_` in your suggestion) is assigned all of the remaining items.  I was hoping for a way to retain that information for future readers/copiers in case they attempted to make adjustments that wouldn't work as they expect.

wlach replied:
Ah ok, you learn something new every day. ;) I might use an underscore just for the first element then. But as I said above, not a big deal.

picnixz replied:
> and yet the last value (the second _ in your suggestion) is assigned all of the remaining items

Technically, no. It would be assigned the remaining items if you were using `...rest`. You can check:

```javascript
[_, two, _] = [1,2,3,4,5]
console.log(`${_}`) // 3

[_, two, ..._] = [1,2,3,4,5]
console.log(`${_}`) // 3,4,5
```

It's the same as in Python:

```python
(one, two, *threefourfive) = (1, 2, 3, 4, 5)
```

jayaddison replied:
Ok, yep - I misunderstood JavaScript's unpacking behaviour.  I'll adjust this to use meaningful-and-underscore-prefixed variable names, and with unpacking for the remainder items.

wlach commented on tests/test_search.py:
```diff
@@ -346,3 +349,20 @@ def assert_is_sorted(item, path: str):
+    app.build()
+
+    fresh_searchindex = (app.outdir / 'searchindex.js')
+    fresh_hash = sha256(fresh_searchindex.read_bytes()).digest()
```
It's a bit unclear to me why you're taking a hash of the file rather than just comparing the contents directly? Feels like that would be just as fast and probably simpler given the sizes of our files.

jayaddison replied:
Initially I felt some concern about this code perhaps being reused for larger files -- to the extent that entire-file comparison might not be ideal -- however, it is true that the search indexes are small, and so the simplicity benefit of comparing the full content does seem worthwhile.

wlach reviewed (commented):
Nice work! I left a few comments which are mostly just nits.

jayaddison committed:
Fixup: relocate CHANGES entry for this feature to in-development section.

jayaddison committed:
Fixup: add self-credit to CHANGES entry.

jayaddison committed:
Tests: attempt to add explanatory content to JS test fixture source projects.

jayaddison committed:
Tests: regenerate JavaScript fixtures by running 'python utils/generate_js_fixtures.py'

jayaddison committed:
Tests: compare regenerated-vs-existing search indices byte-for-byte instead of comparing their sha256 hash values.

wlach reviewed (approved):
Looks good from my perspective! Obviously you'll need a committer to approve/land.

picnixz commented on utils/generate_js_fixtures.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def build(srcdir: Path) -> None:
+    cmd = ('sphinx-build', '-E', '-q', '-b', 'html', f'{srcdir}', f'{srcdir}/_build')
```
Can you use full options here ? It'd be a bit clearer in general (I think we use full options in CI/CD so that's why).

picnixz commented on tests/test_search.py:
```diff
@@ -346,3 +348,17 @@ def assert_is_sorted(item, path: str):
+
+@pytest.mark.parametrize('directory', (
+    directory for directory in (TESTS_ROOT / 'js' / 'roots').iterdir()
+))
```
```suggestion
JAVASCRIPT_TEST_ROOTS = list((TESTS_ROOT / 'js' / 'roots').iterdir())
@pytest.mark.parametrize('directory', JAVASCRIPT_TEST_ROOTS)
```

For parametrized tests, I prefer having a non-iterator to ensure that we have everything *before* launching the test.

jayaddison replied:
Although it may be non-obvious, this should evaluate and construct a non-iterator value for `JAVASCRIPT_TEST_ROOTS` -- the `iterdir()` call is wrapped within a `list(...)` call that should evaluate it.  (or maybe I misunderstood the suggestion here?)

picnixz replied:
Ma suggestion is to have a list in memory when calling `parametrize` and not a generator constructed on the fly. Here,

```
(directory for directory in (TESTS_ROOT / 'js' / 'roots').iterdir())
``` 

is a generator (not a tuple!) (so it is still lazily evaluated) and actually equivalent to `(TESTS_ROOT / 'js' / 'roots').iterdir()`. In general, pytest recommends to have eager values and not iterated values (for instance, if you use my suggestion, it's easy to overwrite the value of JAVASCRIPT_TEST_ROOTS).

jayaddison replied:
Argh, my mistake.  Somehow I read your code review suggestion (that _does_ include a `list(...)` call) as the current state of the test code.  However, indeed, as-is this branch does use an iterator.  Fixing this momentarily..

picnixz commented on tests/js/fixtures/partial/searchindex.js:
```diff
@@ -0,0 +1 @@
+Search.setIndex({"alltitles": {"sphinx_utils module": [[0, "sphinx-utils-module"]]}, "docnames": ["index"], "envversion": {"sphinx": 61, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains ...
```
I don't remember whether I told/asked you or not, but do we need a blank line at the end of this file (and the others)?

![image](https://github.com/sphinx-doc/sphinx/assets/10796600/07718a55-021a-4570-ae12-b8abd45bc072)

jayaddison replied:
As output by Sphinx itself, the `searchindex.js` file does not include an end-of-file newline, and since the fixtures are generated by Sphinx itself, I'd prefer that they don't contain those either.

At one point I did dabble with adding auto-reformatting of the fixtures -- arguably it could help when reviewing diffs that affect the index structure -- but on balance I preferred the simplicity of not adding any post-processing steps (and dependencies) for the JS file (adding some emphasis there: the index file is JavaScript code, not JSON data).

jayaddison committed:
[tests] search: code review: add variable names for parsed-but-unused query terms items.

jayaddison committed:
[utils] js fixtures: code review: use fully-expanded option names when invoking sphinx command.

jayaddison committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
[utils] js fixtures: lint fixup: use argument expansion to populate builder option.

picnixz reviewed (approved):
For me it's fine now (modulo this small suggestion but we can just decide later).

picnixz committed:
Merge branch 'master' into issue-12099-prep/extract-js-search-index-fixtures

jayaddison committed:
[tests] search: provide eager-evaluated JavaScript test roots instead of reading them from an iterator at test-time.

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

jayaddison committed:
[tests] search: relocate constant declaration to appease linters.
